qcelemental/models/basis.py

```diff
@@ -95,18 +110,58 @@ class BasisSet(ProtoModel):
     schema_name: constr(strip_whitespace=True, regex="qcschema_basis") = "qcschema_basis"
     schema_version: int = 1
 
-    basis_name: str = Schema(..., description="A standard basis name if available (e.g., 'cc-pVDZ'.")
+    name: str = Schema(..., description="A standard basis name if available (e.g., 'cc-pVDZ'.")
     description: Optional[str] = Schema(None, description="A brief description of the basis set.")
-    basis_data: Dict[str, BasisCenter] = Schema(..., description="A mapping of all types of centers available.")
-    basis_atom_map: List[str] = Schema(
-        ..., description="Mapping of all centers in the parent molecule to centers in `basis_data`.")
+    center_data: Dict[str, BasisCenter] = Schema(..., description="A mapping of all types of centers available.")
+    atom_map: List[str] = Schema(
+        ..., description="Mapping of all centers in the parent molecule to centers in `center_data`.")
+
+    nbf: Optional[int] = Schema(None, description="The number of basis functions.")
 
-    @validator('basis_atom_map', whole=True)
+    @validator('atom_map', whole=True)
     def _check_atom_map(cls, v, values):
         sv = set(v)
-        missing = sv - values["basis_data"].keys()
+        missing = sv - values["center_data"].keys()
 
         if missing:
-            raise ValueError(f"'basis_atom_map' contains unknown keys to 'basis_data': {missing}.")
+            raise ValueError(f"'atom_map' contains unknown keys to 'center_data': {missing}.")
+
+        return v
+
+    @validator('nbf', always=True)
+    def _check_nbf(cls, v, values):
+
+        # Bad construction, pass on errors
+        try:
+            nbf = cls._calculate_nbf(values["atom_map"], values["center_data"])
+        except KeyError:
+            return v
+
+        if v is None:
+            v = nbf
+        else:
+            if v != nbf:
+                raise ValidationError("Calculated nbf does not match supplied nbf.")
```

Comment:
should user _ever_ be supplying nbf? I'd think it'd always be on-the-fly. Or is this for a revalidation check dumping in a pre-built object that's been living loose as a dictionary for a while?

Reply:
It is hard to not allow a variable in Pydantic, but ensure it is always a value. This is my awkward solution to this problem.

Reply:
ah, you want `bsinstance.nbf` valid rather than `bsinstance.nbf()` for otf? sounds reasonable